nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/ui/provenance-event-listing/provenance-event-table/lineage/lineage.component.ts

```diff
@@ -0,0 +1,975 @@
+                    // TODO - cluster node id
+                    self.submitLineageQuery.next({
+                        lineageRequestType: 'PARENTS',
+                        eventId: selectionData.id
+                        // clusterNodeId: clusterNodeId
+                    });
+                }
+            },
+            {
+                condition: function (selection: any, self: LineageComponent) {
+                    if (selection.empty()) {
+                        return false;
+                    }
+
+                    const selectionData: any = selection.datum();
+                    return self.supportsExpandCollapse(selectionData);
+                },
+                clazz: 'fa fa-plus-square',
+                text: 'Expand',
+                action: function (selection: any, self: LineageComponent) {
+                    const selectionData: any = selection.datum();
+
+                    // TODO - cluster node id
+                    self.submitLineageQuery.next({
+                        lineageRequestType: 'CHILDREN',
+                        eventId: selectionData.id
+                        // clusterNodeId: clusterNodeId
+                    });
+                }
+            },
+            {
+                condition: function (selection: any, self: LineageComponent) {
+                    if (selection.empty()) {
+                        return false;
+                    }
+
+                    const selectionData: any = selection.datum();
+                    return self.supportsExpandCollapse(selectionData);
+                },
+                clazz: 'fa fa-minus-square',
+                text: 'Collapse',
+                action: function (selection: any, self: LineageComponent) {
+                    const selectionData: any = selection.datum();
+                    self.collapseLineage(selectionData);
+                }
+            }
+        ]
+    };
+
+    private allMenus: Map<string, ContextMenuDefinition>;
+
+    lineageElement: any;
+    lineageContainerElement: any;
+    lineageContextmenu: ContextMenuDefinitionProvider;
+
+    private nodeLookup: Map<string, any> = new Map<string, any>();
+    private linkLookup: Map<string, any> = new Map<string, any>();
+    private previousEventTimestampThreshold: number = -1;
+
+    constructor() {
+        this.allMenus = new Map<string, ContextMenuDefinition>();
+        this.allMenus.set(this.ROOT_MENU.id, this.ROOT_MENU);
+
+        const self: LineageComponent = this;
+        this.lineageContextmenu = {
+            getMenu(menuId: string): ContextMenuDefinition | undefined {
+                return self.allMenus.get(menuId);
+            },
+            filterMenuItem(menuItem: ContextMenuItemDefinition): boolean {
+                // include if the condition matches
+                if (menuItem.condition) {
+                    const selection: any = d3.select('circle.context');
+                    return menuItem.condition(selection, self);
```

Comment:
Assuming you take the suggestion about not requiring `self` not needing to be passed in, you can remove it as an argument here.